hivemind/utils/mpfuture.py

```diff
@@ -1,147 +1,164 @@
+        self._sync_updates()
+        return self._state == base.RUNNING
 
     def cancelled(self):
-        warn("cancelled not implemented")
-        return False
+        self._sync_updates()
+        return self._state == base.CANCELLED
 
     def add_done_callback(self, callback):
-        raise NotImplementedError()
-
-    def __repr__(self):
-        try:
-            self._recv(timeout=0)
-        except TimeoutError:
-            pass
-        if self.state == self.STATE_FINISHED:
-            return "<MPFuture at 0x{:x} state=finished returned {}>".format(id(self), type(self._result))
-        elif self.state == self.STATE_EXCEPTION:
-            return "<MPFuture at 0x{:x} state=finished raised {}>".format(id(self), type(self._exception))
-        else:
-            return "<MPFuture at 0x{:x} state={}>".format(id(self), self.state)
+        raise NotImplementedError(f"MPFuture doesn't support callbacks.")
 
-    async def _async_recv(self, timeout):
-        loop = asyncio.get_running_loop()
+    def remove_done_callback(self, callback):
+        raise NotImplementedError(f"MPFuture doesn't support callbacks.")
 
-        if self.state in (self.STATE_PENDING, self.STATE_RUNNING):
-            status, payload = await loop.run_in_executor(None, self.poll_and_recv, timeout)
+    def get_loop(self):
+        raise NotImplementedError(f"MPFuture doesn't support get_loop")
 
-            assert status in self.STATES
-            self.state = status
+    @property
+    @lru_cache()
+    def _shutdown_trigger(self):
+        return base.Future()
 
-            if status == self.STATE_FINISHED:
-                self._result = payload
-            elif status == self.STATE_EXCEPTION:
-                self._exception = payload
-            elif status in (self.STATE_RUNNING, self.STATE_CANCELLED):
-                pass  # only update self.state
+    def __repr__(self):
+        self._sync_updates()
+        if self._state == base.FINISHED:
+            if self._exception:
+                return "<MPFuture at 0x{:x} state=finished raised {}>".format(id(self), type(self._exception))
             else:
-                raise ValueError("Result status should not be self.STATE_PENDING")
+                return "<MPFuture at 0x{:x} state=finished returned {}>".format(id(self), type(self._result))
+        else:
+            return "<MPFuture at 0x{:x} state={}>".format(id(self), self._state)
 
-    async def async_result(self, timeout=None):
-        await self._async_recv(timeout)
-        if self.state == self.STATE_FINISHED:
-            return self._result
-        elif self.state == self.STATE_EXCEPTION:
+    def __await__(self):
+        yield from asyncio.get_running_loop().run_in_executor(None, self._await_terminal_state, None).__await__()
+        if self._exception:
             raise self._exception
-        else:
-            assert self.state == self.STATE_CANCELLED
-            raise CancelledError()
+        return self._result
```

Comment:
changed method name to match asyncio.Future

/* asyncio.Future itself is awaitable while .result / .exception return current values (if any) */